radixdlt/src/main/java/com/radixdlt/consensus/sync/SyncedRadixEngine.java

```diff
@@ -0,0 +1,173 @@
+ */
+@Singleton
+public class SyncedRadixEngine implements SyncedStateComputer<CommittedAtom> {
+	private static final Logger log = LogManager.getLogger();
+	private final RadixEngine<LedgerAtom> radixEngine;
+	private final CommittedAtomsStore committedAtomsStore;
+	private final AddressBook addressBook;
+	private final StateSyncNetwork stateSyncNetwork;
+
+	@Inject
+	public SyncedRadixEngine(
+		RadixEngine<LedgerAtom> radixEngine,
+		CommittedAtomsStore committedAtomsStore,
+		AddressBook addressBook,
+		StateSyncNetwork stateSyncNetwork
+	) {
+		this.radixEngine = radixEngine;
+		this.committedAtomsStore = committedAtomsStore;
+		this.addressBook = addressBook;
+		this.stateSyncNetwork = stateSyncNetwork;
+	}
+
+	/**
+	 * Start the service
+	 */
+	public void start() {
+		stateSyncNetwork.syncRequests()
+			.observeOn(Schedulers.io())
+			.subscribe(syncRequest -> {
+				log.info("SYNC_REQUEST: {} {}", syncRequest, this.committedAtomsStore.getStateVersion());
+				Peer peer = syncRequest.getPeer();
+				long stateVersion = syncRequest.getStateVersion();
+				// TODO: This may still return an empty list as we still count state versions for atoms which
+				// TODO: never make it into the radix engine due to state errors. This is because we only check
+				// TODO: validity on commit rather than on proposal/prepare.
+				// TODO: remove 100 hardcode limit
+				List<CommittedAtom> committedAtoms = committedAtomsStore.getCommittedAtoms(stateVersion, 100);
+				log.info("SYNC_REQUEST: SENDING_RESPONSE {}", committedAtoms);
+				stateSyncNetwork.sendSyncResponse(peer, committedAtoms);
+			});
+
+		stateSyncNetwork.syncResponses()
+			.observeOn(Schedulers.io())
+			.subscribe(syncResponse -> {
+				// TODO: Check validity of response
+				log.info("SYNC_RESPONSE: {}", syncResponse);
+				for (CommittedAtom committedAtom : syncResponse) {
+					if (committedAtom.getVertexMetadata().getStateVersion() > this.committedAtomsStore.getStateVersion()) {
+						this.execute(committedAtom);
+					}
+				}
+			});
+	}
+
+	/**
+	 * Initiate a sync to a target state version. Searches for a peer
+	 * given a target list and then requests for atom inventory.
+	 *
+	 * TODO: cancel previous unfinished requests
+	 *
+	 * @param targetStateVersion the target state version
+	 * @param target a list of potential targets
+	 * @return completable which completes when the store has completed syncing
+	 */
+	@Override
+	public Completable syncTo(long targetStateVersion, List<ECPublicKey> target) {
+		final long currentStateVersion = committedAtomsStore.getStateVersion();
+		if (targetStateVersion <= currentStateVersion) {
+			return Completable.complete();
+		}
+
+		Peer peer = target.stream()
+			.map(pk -> addressBook.peer(pk.euid()))
+			.filter(Optional::isPresent)
+			.map(Optional::get)
+			.filter(Peer::hasSystem)
+			.findFirst()
+			.orElseThrow(() -> new RuntimeException("Unable to find peer"));
+		stateSyncNetwork.sendSyncRequest(peer, currentStateVersion);
```

Comment:
If a node is lagging behind a bit, it may issue many requests to whoever the first viable peer in `target` is before actually catching up (because requests don't coalesce). Is this something we want to change later or do you think it's not an issue? 
If we want to come back to it later, we may at least want to randomly pick our peer from `target`.

Reply:
yeah definitely we want to make better for future